src/quota_aggregator_impl.cc

```diff
@@ -0,0 +1,262 @@
+    const ::google::api::servicecontrol::v1::AllocateQuotaRequest& request,
+    const ::google::api::servicecontrol::v1::AllocateQuotaResponse& response) {
+  if (!cache_) {
+    return ::google::protobuf::util::Status::OK;
+  }
+
+  string request_signature = GenerateAllocateQuotaRequestSignature(request);
+
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer stack_buffer(this);
+  MutexLock lock(cache_mutex_);
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer::Swapper swapper(
+      this, &stack_buffer);
+
+  CacheElem* cache_elem = nullptr;
+
+  if (response.allocate_errors_size() == 0) {
+    // If the current response is positive and the signature is already in the
+    // cache, aggregation is not required. Tokens are already used to allocated
+    // quota for the current request.
+    QuotaCache::ScopedLookup lookup(cache_.get(), request_signature);
+    if (!lookup.Found()) {
+      cache_elem = new CacheElem(response);
+    } else {
+      lookup.value()->set_quota_response(response);
+    }
+  } else {
+    // If the response is negative, insert new or replace existing element from
+    // the current response.
+    cache_elem = new CacheElem(response);
+  }
+
+  if (cache_elem != nullptr) {
+    cache_elem->set_signature(request_signature);
+    cache_->Insert(request_signature, cache_elem, 1);
+  }
+
+  return ::google::protobuf::util::Status::OK;
+}
+
+// When the next Flush() should be called.
+// Returns in ms from now, or -1 for never
+int QuotaAggregatorImpl::GetNextFlushInterval() {
+  if (!cache_) return -1;
+  return options_.refresh_interval_ms;
+}
+
+// Invalidates expired allocate quota responses.
+// Called at time specified by GetNextFlushInterval().
+::google::protobuf::util::Status QuotaAggregatorImpl::Flush() {
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer stack_buffer(this);
+  MutexLock lock(cache_mutex_);
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer::Swapper swapper(
+      this, &stack_buffer);
+
+  if (cache_) {
+    cache_->RemoveExpiredEntries();
+  }
+
+  return Status::OK;
+}
+
+// Flushes out all cached check responses; clears all cache items.
+// Usually called at destructor.
+::google::protobuf::util::Status QuotaAggregatorImpl::FlushAll() {
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer stack_buffer(this);
+  MutexLock lock(cache_mutex_);
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer::Swapper swapper(
+      this, &stack_buffer);
+
+  is_refresh_stop_ = true;
+
+  GOOGLE_LOG(INFO) << "Remove all entries of check aggregator.";
+  if (cache_) {
+    cache_->RemoveAll();
+  }
+
+  return Status::OK;
+}
+
+// OnCacheEntryDelete will be called behind the cache_mutex_
+// no need to consider locking at this point
+//
+// if the element is refreshing and aggregated while waiting for the
+// response, tokens should be aggregated
+void QuotaAggregatorImpl::OnCacheEntryDelete(CacheElem* elem) {
+  if (elem->is_aggregated() == true && is_refresh_stop_ == false) {
```

Comment:
Can we ignore aggregated tokens before the response arrives?
I think if the response is negative, we still need to allocate quota for those tokens.